src/controller.js

```diff
@@ -0,0 +1,601 @@
+        service: 's3',
+        region: this.region,
+        method: 'PUT',
+        protocol: this.s3protocol,
+        hostname: this.__s3hostname(bucket),
+        path: `/${key}?partNumber=${num}&uploadId=${uploadId}`,
+        headers: {
+          'x-amz-content-sha256': part.sha256,
+          'content-length': part.size,
+        }
+      });
+
+      requests.push(this.__serializeRequest(signedRequest));
+    }
+
+    return requests;
+  }
+
+  /**
+   * This method is used to tag resources, and only after completion.  This is
+   * done to maintain parity with single part uploads and should *not* be
+   * relied upon to be atomic.  This is designed to be used in things like the
+   * cost explorer.  It is not intended to be part of the public api of this
+   * library, hence the name.  If it were possible to tag the multipart upload
+   * at object creation as it is with single part uploads, we'd do that instead
+   *
+   * http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUTtagging.html
+   */
+  async __tagObject(opts) {
+    opts = runSchema(opts, Joi.object().keys({
+      bucket: schemas.bucket.required(),
+      key: schemas.key.required(),
+      tags: schemas.tags.required(),
+    }));
+
+    let {bucket, key, tags} = opts;
+
+    let requestBody = this.__generateTagSetBody(tags);
+
+    let signedRequest = aws4.sign({
+      service: 's3',
+      region: this.region,
+      method: 'PUT',
+      protocol: this.s3protocol,
+      hostname: this.__s3hostname(bucket),
+      path: `/${key}?tagging=`,
+      body: requestBody,
+    });
+
+    let response = await this.runner({
+      req: this.__serializeRequest(signedRequest),
+      body: requestBody,
+    });
+
+
+    parseS3Response(response.body);
+
+    if (response.statusCode !== 200) {
+      throw new Error('Could not tag object');
+    }
+  }
+
+  /**
+   * Mark a multipart upload as completed. 
+   *
+   * http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadComplete.html
+   */
+  async completeMultipartUpload(opts) {
+    opts = runSchema(opts, Joi.object({
+      bucket: schemas.bucket.required(),
+      key: schemas.key.required(),
+      etags: schemas.etags.required(),
+      tags: schemas.tags,
+      uploadId: schemas.uploadId.required(),
+    }).optionalKeys('tags'));
+
+    let {bucket, key, uploadId, etags, tags} = opts;
+
+    // I'm not sure why, but for some reason the AWS4 library generates the
+    // incorrect SHA256 for *this* and only *this* body.  I have to calculate
+    // the body sha256 myself, or else what happens is that the requestBody
+    // string is hashed with its value and an extra newline character, but then
+    // the data written does not have that newline, and never did.  I suspect
+    // that there's something broken in the aws4 library here
+    let requestBody = this.__generateCompleteUploadBody(etags);
+    let requestBodySha256 = crypto.createHash('sha256').update(requestBody);
+
+    let signedRequest = aws4.sign({
```

Comment:
There's no reason to use aws-sdk.  The signatures are tested and the XML generation and parsing is tested.  Is there a test case that you think should be added?